firedrake/extrusion_numbering.pyx

```diff
@@ -0,0 +1,762 @@
+the base topological entity, and *above* it in each cell::
+
+       x-------x
+       |   o   |
+       o   o   o
+       o   o   o
+       |   o   |
+       o---o---o
+
+This element has one degree of freedom on each base vertex and cell,
+two degrees of freedom "above" each vertex, and four above each cell.
+To compute the number of degrees of freedom on the column of
+topological entities we sum the number on the entity, multiplied by
+the number of layers with the number above, multiplied by the number
+of layers minus one (due to the fencepost error difference).
+This number of layers naturally changes from entity to entity, and so
+we can't compute this up front, but must do it point by point,
+constructing the PETSc Section as we go.
+
+Computation of function space maps
+----------------------------------
+
+Now we need the maps from topological entities (cells and facets) to
+the function space nodes they can see.  The allocation offsets that
+the numbering section gives us are wrong, because when we have a step
+in the column height, the offset will be wrong if we're looking from
+the higher cell.  Consider a vertex discretisation on the previous
+mesh, with a numbering::
+
+                      8--------10
+                      |        |
+                      |        |
+                      |        |
+                      |        |
+    2--------5--------7--------9
+    |        |        |
+    |        |        |
+    |        |        |
+    |        |        |
+    1--------4--------6
+    |        |
+    |        |
+    |        |
+    |        |
+    0--------3
+
+The cell node map we get by just looking at allocation offsets is::
+
+   [[0, 1, 3, 4],
+    [3, 4, 6, 7],
+    [6, 7, 9, 10]]
+
+note how the second and third cells have the wrong value for their
+"left" vertices.  Instead, we need to shift the numbering we obtain
+from the allocation offset by the number of entities we're skipping
+over, to result in::
+
+   [[0, 1, 3, 4],
+    [4, 5, 6, 7],
+    [7, 8, 9, 10]]
+
+Now, when we iterate over cells, we ensure that we access the correct
+dofs.  The same trick needs to be applied to facet maps too.
+
+Computation of boundary nodes
+-----------------------------
+
+For the top and bottom boundary nodes, we walk over the cells at,
+respectively, the top and bottom of the column and pull out those
+nodes whose entity height matches the appropriate cell height.  As an
+example::
+
+                      8--------10
+                      |        |
+                      |        |
+                      |        |
+                      |        |
+    2--------5--------7--------9
+    |        |        |
+    |        |        |
+    |        |        |
+    |        |        |
+    1--------4--------6
+    |        |
+    |        |
+    |        |
+    |        |
+    0--------3
+
+The bottom boundary nodes are::
+
+   [0, 3, 6, 9]
+
+whereas the top are::
+
+   [2, 5, 8, 10]
```

Comment:
>On an unstructured mesh I'd expect the "bottom" boundary nodes to be `[0, 3, 4, 6, 7, 9]` and the "top" to be `[2, 5, 7, 8, 10]`.

I agree with you. My concern is how should we handle nodes that are, sort of, both "bottom" and "top" (such as node 7). Will this be a problem?